Generic model support

aryanagarwal9 opened the pull request:
* **Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes/features)
- [x] Docs have been added / updated (for bug fixes / features)


* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
feature

* **What is the current behavior?** (You can also link to an open issue here)
Only models specified in the model_registry can be used

* **What is the new behavior (if this is a feature change)?**
generic models, passed as a dictionary, are supported

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
No

* **Other information**:

aryanagarwal9 committed:
Create test_generic_hf.py

aryanagarwal9 committed:
Added generic hf model support

aryanagarwal9 committed:
Temporarily removed a model for testing purpose

aryanagarwal9 committed:
turned back into original version

aryanagarwal9 committed:
Delete test_generic_hf.py

aryanagarwal9 committed:
Merge branch 'mainline' into generic-hf-support

aryanagarwal9 committed:
Added s2_inference errors

aryanagarwal9 committed:
generic model support error handling

aryanagarwal9 committed:
added test_generic_model_support

aryanagarwal9 committed:
Refactor

aryanagarwal9 committed:
added comments

aryanagarwal9 committed:
added error message

aryanagarwal9 committed:
changed name of test file

aryanagarwal9 committed:
_check_model_dict is called earlier

aryanagarwal9 committed:
pass model properties to vectorise

aryanagarwal9 committed:
add model_property validation

aryanagarwal9 committed:
error handling

aryanagarwal9 committed:
addede tests

aryanagarwal9 committed:
shifted error to s2_inference errors

aryanagarwal9 committed:
updated error handling

aryanagarwal9 committed:
Update test_generic_model.py

aryanagarwal9 committed:
Update test_generic_model.py

aryanagarwal9 committed:
Merge branch 'mainline' into generic-hf-support

aryanagarwal9 committed:
Merge branch 'generic-hf-support' of https://github.com/marqo-ai/marqo into generic-hf-support

aryanagarwal9 committed:
Refactored getting model_properties

aryanagarwal9 committed:
changed cache key and added test

aryanagarwal9 committed:
Update errors.py

aryanagarwal9 committed:
error renaming

aryanagarwal9 committed:
refactor error name

aryanagarwal9 committed:
fixed model_properties bug

aryanagarwal9 committed:
updated tests

aryanagarwal9 committed:
Merge branch 'generic-model-support-2' into generic-hf-support

pandu-k committed:
Merge pull request #178 from marqo-ai/generic-hf-support

Generic hf support to generic model support

pandu-k commented on src/marqo/s2_inference/s2_inference.py:
```diff
@@ -13,12 +15,13 @@
     Args:
-        model_name (str): _description_
+        model_name (str) : _description_
         content (_type_): _description_
```
Please add description for model_properties

aryanagarwal9 committed:
Merge branch 'mainline' into generic-model-support-2

aryanagarwal9 committed:
fix tests

aryanagarwal9 committed:
Merge branch 'mainline' into generic-model-support-2

aryanagarwal9 committed:
added docs

aryanagarwal9 committed:
fixed encoding tests

aryanagarwal9 committed:
fixed cache_key tests

modified _create_cache_key function
and validating model properties before calling _create_model_cache

aryanagarwal9 committed:
fixed tests

gazambuja commented:
@jn2clark any ETA for this?

jn2clark commented:
@gazambuja Monday 12 Dec is the scheduled next release

aryanagarwal9 committed:
Merge branch 'mainline' into generic-model-support-2

aryanagarwal9 committed:
Refactored  _create_model_cache_key

Changed the function signature of _create_model_cache_key

aryanagarwal9 committed:
Refactored comments

aryanagarwal9 committed:
Optimised imports

aryanagarwal9 committed:
Refactored _create_model_cache_key

aryanagarwal9 committed:
Merge branch 'mainline' into generic-model-support-2

pandu-k committed:
Catching s2_inference vectorise errors in search

pandu-k commented:
I ran a manual test, just using a HF model name (no model properties) There are two problems I ran into:

1. It allowed me to index all the documents, but on search an error is raised. If we can't search an index, it should prevent us indexing documents in it. A helpful error should be raised.
2. The error message does not propagate nicely to the user. It is a 500 error, without much information. I've made [this branch](https://github.com/marqo-ai/marqo/tree/catching_search_inference_errors) (feel free to merge it into this PR) which should at least help the error message propagate to the user. We need to make the error message even more helpful, such as how the user can fix the error.

Update: this error pattern also occurs when the model name is misspelled.

Code to reproduce problem:
```
mq = marqo.Client()
index_name = 'my-test-index'
model_name = 'sentence-transformers/paraphrase-albert-small-v2'

def set_up():
  try:
    mq.delete_index(index_name)
  except marqo.errors.IndexNotFoundError as e:
    pass


def generate_data(num_docs = 1000):
  vocab_source = "https://www.mit.edu/~ecprice/wordlist.10000"

  vocab = requests.get(vocab_source).text.splitlines()

  docs = [{"Title": " ".join(random.choices(population=vocab, k=10)),
           "Description": " ".join(random.choices(population=vocab, k=25)),
           "_id": f"doc_{i}"
           } for i in range(num_docs)]
  return docs


def random_1k_test():
  set_up()
  ## Set up index:
  mq.create_index(index_name=index_name, model=model_name)

  data = generate_data(num_docs=1000)
  mq.index(index_name).add_documents(data, client_batch_size=20)

  ix = mq.index(index_name)

  # ERROR OCCURS HERE:
  res = ix.search("Some string")

  top_id = res['hits'][0]['_id']
  print('results')
  pprint.pprint(res)
  print('get facet')
  pprint.pprint(ix.get_document(top_id, expose_facets=True))


random_1k_test()
```
The error that is received is an internal error. To the client it is unhelpful:
```
marqo.errors.MarqoWebError: MarqoWebError: MarqoWebError Error message: Internal Server Error
status_code: 500, type: unhandled_error_type, code: unhandled_error, link: 
```
The logs are more helpful:

```
marqo.s2_inference.errors.UnknownModelError: Could not find model properties for model=sentence-transformers/paraphrase-albert-small-v2
```

aryanagarwal9 committed:
Merge remote-tracking branch 'origin/catching_search_inference_errors' into generic-model-support-2

aryanagarwal9 committed:
Merge branch 'mainline' into generic-model-support-2

aryanagarwal9 committed:
Throw useful errors

Better error thrown if only model_name is given but it is not in registry and model_properties are missing

aryanagarwal9 committed:
Throw useful errors

Added better errors if:

1. user puts in model_name and model_properties but does not give the required keys in model_properties.

2. user puts in model_name, model_properties, gives all the required key, but gives the wrong model name in model_properties['name']

aryanagarwal9 committed:
added test

aryanagarwal9 committed:
added better error messages

aryanagarwal9 committed:
refactored logs

aryanagarwal9 committed:
code style changes

aryanagarwal9 committed:
added test

aryanagarwal9 committed:
Merge branch 'mainline' into generic-model-support-2

pandu-k committed:
Delete .DS_Store

pandu-k committed:
Merge branch 'mainline' into generic-model-support-2

pandu-k commented:
Merged mainline in again.
- Passed unit tests
- Failed DIND API tests, but all failed tests corresponded to those introduced in the visual search update

pandu-k commented:
Squash merged back into mainline. Running all CI tests.
- local OS passed (besides VS tests)
- S2Search Passed (besides VS tests)
- DIND OS Passed (besides VS tests)
- Cuda DIND OS Passed (besides VS tests)
- Arm64 Local OS Passed (besides VS tests)
- Unit tests all passed
**Note:**
- VS tests == visual search API tests
